addons/web_editor/static/src/js/editor/snippets.editor.js

```diff
@@ -1099,21 +1106,40 @@ var SnippetEditor = Widget.extend({
-                    if (isPopup && !$selectorSiblings[i].closest('div.s_popup')) {
-                        // Removing the siblings that are outside the popup if
-                        // the grid item is in a popup.
+        if (rowEl.classList.contains('row') && $selectorSiblings) {
+            // Getting the closest grid of the column at the start of the drag
+            // (may be undefined, if it cannot toggle the grid mode).
+            const columnGridEl = rowEl.closest('.o_grid_mode');
+            // Looping backwards because elements are removed, so the
+            // indexes are not lost.
+            for (let i = $selectorSiblings.length - 1; i >= 0; i--) {
+                if (isPopup && !$selectorSiblings[i].closest('div.s_popup')) {
+                    // Removing the siblings that are outside the popup if
+                    // the grid item is in a popup.
+                    $selectorSiblings.splice(i, 1);
+                } else {
+                    const gridSnippetEl = $selectorSiblings[i].closest('.o_grid_mode');
+                    // If we drag a normal column (= not grid), all the siblings
+                    // having a closest grid are removed (because columnGridEl
+                    // is undefined so the condition is true).
+                    // If we drag a grid column (= direct child of grid), every
+                    // siblings having a closest grid are removed because either
+                    // they are in another snippet (gridSnippetEl is different
+                    // than columnGridEl) or their closest grid is the parent of
+                    // the column (gridSnippetEl === rowEl).
+                    // If we drag a column of a snippet that is in a grid column
+                    // (= inner snippet column), the inner siblings are kept
+                    // because they are the only ones having the same closest
+                    // grid as the column and which is not the parent of the
+                    // column. (*)
+                    // *: Note that it works only for non-grid inner snippets
+                    // whose columns cannot be dropped outside them (i.e. form
+                    // snippet). It may need to be adapted when other snippets
+                    // not as restricted as the form could be dropped inside
+                    // columns in a hypothetical future.
+                    if (gridSnippetEl && (gridSnippetEl !== columnGridEl || gridSnippetEl === rowEl)) {
                         $selectorSiblings.splice(i, 1);
-                    } else {
-                        const gridSnippet = $selectorSiblings[i].closest('div.o_grid_mode');
-                        if (gridSnippet) {
-                            $selectorSiblings.splice(i, 1);
-                            selectorGrids.add(gridSnippet);
-                        }
+                        selectorGrids.add(gridSnippetEl);
```

Comment:
1. I don't understand why you are talking about `$selectorChildren` because when we are dragging a column, there are only siblings, as there are only `drop-near` rules for columns (so `$selectorChildren` is undefined). I applied the siblings filtering only to the columns since they are the only ones concerned by grid, which is why there is the check if the parent has the `row` class at the beginning. This same check also prevents the filtering to be applied on the inner contents (alert snippet,...) since they are always inside a column (so the row is never their parent). The same goes for building blocks (since they can't have a parent with a `row` class).

2. I agree that only checking if the parent is a grid is enough, as said in the first review (I was being too cautious when I did `closest` as I wanted to be sure the siblings would be removed but it was stupid). In fact, all this fix is needed only because we can drop columns inside the form (because then we have the form dropzones overlapping with the grid dropzone). So if we can also prevent columns from other snippets to be dropped inside a form, then I think the fix would be to do that and to only check the parent of the the siblings. (Or if you think it doesn't matter that much to have the overlapping dropzones, then only checking the parent is enough).

3. For the popup, I think that there is no need to do that because it is managed in `_activateInsertionZones` (like I said, I should have placed the filtering inside this function after the part about the modals so I will fix this in another PR by moving the code there. If you agree of course). 

What do you think ? :)

Reply:
@sobo-odoo 

> I don't understand why you are talking about $selectorChildren because when we are dragging a column, there are only siblings, as there are only drop-near rules for columns

Yes that's my point :wink:  -> What if someone defines drop-in rules which end up allowing things in grid areas? Odoo code can evolve, it should be robust to that evolution. But even more importantly, Odoo can be customized, meaning someone can just says "hey my new snippet portion here can be dropped *inside* (drop-in) rows of text snippets"... not thinking of grid and your code will not work because that external developer used a drop-in rule :wink: 

We should have done the same for the s_popup / table of content / form snippet above (considering there might be drop-in rules) but that's not as important as the grid code since it's generic to all snippets.

> all this fix is needed only because we can drop columns inside the form (because then we have the form dropzones overlapping with the grid dropzone)

Indeed, now that I re-see it I think it can make sense to want dragging a column inside of a form. But as this behavior is possible from any column to any form, it's strange we can do it when the form is inside another snippet in columns mode but cannot do it when the form is inside another snippet in grid mode. I don't know... maybe it's not really bad to allow having overlapping "normal dropzones" with "grid dropzone" for this really specific case:
- It's consistent with the normal behavior allowing dropping items in-between form fields
- In the end, it's ok to use I think: you can drag elements in the grid and once you hover the form, you switch to dropping inside the form, which is ok I think because:
    - Why would you want to put something over your form?
    - Very corner case to have a snippet in grid mode + a form dropped inside (I would already discourage users doing that IMO) => not worth any code to change the behavior IMO
    
> For the popup, I think that there is no need to do that because it is managed in _activateInsertionZones

Not important for 16.0. But in master, we should find a way to unify that logic of "filtering out dropzones", it should probably be included directly in the functions that find `$selectorSiblings` and `$selectorChildren` in the first place (so neither here nor `_activateInsertionZones` IMO).

Reply:
Okok thanks a lot for the explanation :) (sorry :sweat_smile:)
(Just to recapitulate: 
- I will do like you said for the siblings/children
- I won't touch anything about the form 
- I will filter the popup like you said here:
> to handle point 5, it should be done just like that too...
- The rest is for another PR in master)

I just have one last thing to say about the form: 
> Why would you want to put something over your form?

Not over but you might want to put something behind it (like a colored column to have a special background)
> Very corner case to have a snippet in grid mode + a form dropped inside

If you go to the "Contact Us" page, you have a form dropped in a Text snippet. You just have to drag a column to be in this situation so I don't think it is that much of a corner case. 

Knowing that, is it still alright to not do anything about the form ?

Reply:
> Not over but you might want to put something behind it (like a colored column to have a special background)

That's true. I doubt it's that useful but it will still be a possibility anyway:
- Move the form elsewhere
- Put you colored column behind
- Reput your form on top of it

Not ideal way to achieve that but not ideal to try and put things behind your form too IMO :grimacing: 

> If you go to the "Contact Us" page, you have a form dropped in a Text snippet. You just have to drag a column to be in this situation so I don't think it is that much of a corner case.

That's true too but I doubt it will be a problem really. I really don't see someone complaining with "I cannot put something over/under my form, it is automatically nicely put in between my fields" :grimacing: But maybe... :)

> Knowing that, is it still alright to not do anything about the form ?

Let's ask @pierre-pde 's opinion, he can decide:
A. We can drop any column between form fields. Should we keep that feature?
B. If yes, when the form is inside another snippet's column and that snippet is in grid mode:
     1. Should we disable A as the normal dropzones overlaps with the grid mode dropzones?
     2. If yes, should we disable A for all form used inside other snippets (whether or not they are grid mode) to be consistent or not?

My opinion is:
- A -> Yes, that seems like a good enough feature. For example, here is "why not" case where I added my social media next to a form field:
![image](https://user-images.githubusercontent.com/10338094/211880736-7719ee21-a282-4258-affa-22352570bc17.png)
- B.2 -> Contact us page is exactly that case, I would really like to keep A in that case too (because why not?)
- B.1 -> I don't like inconsistencies, if we can drop things in between form fields, I would really like that to be the case everywhere. So I would keep dropzones overlapping with grid mode to stay consistent.

But any opinion is fine really, tell us what you think :+1:

Reply:
Hi @sobo-odoo , @qsm-odoo took the time to 'demo' me the choice.
In the end, I agree with him:
Even if a form is inside a snippet in grid mode, we should keep form normal dropzone.
Which means you might have normal dropzones on top of a grid. 
We can discuss it if I'm not clear, it's hard to describe :D

Reply:
Alright, so I don't change anything about the form
Thanks ! :)